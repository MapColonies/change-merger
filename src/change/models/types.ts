import { OsmChange } from '@map-colonies/node-osm-elements';

interface IdMapping {
  externalId: string;
  tempOsmId: number;
}

interface OsmXmlTag {
  k: string;
  v: string;
}

interface ChangeWithMetadata {
  externalId: string;
  change: OsmChange;
  tempOsmId?: number;
  action: 'create' | 'modify' | 'delete';
}

interface OsmXmlBase {
  id: number;
  version: number;
  changeset: number;
  tag: OsmXmlTag[];
}

interface OsmXmlNode extends OsmXmlBase {
  lon: number;
  lat: number;
}

interface OsmXmlWay extends OsmXmlBase {
  nd: { ref: number }[];
}

interface ChangeActionObj {
  node: OsmXmlNode[];
  way: OsmXmlWay[];
}

 interface OsmXmlChange {
  version: '0.6';
  generator: string;
  create: ChangeActionObj;
  modify: ChangeActionObj;
  delete: ChangeActionObj;
}


export { OsmXmlNode, OsmXmlWay, OsmXmlTag, OsmXmlChange, ChangeWithMetadata, IdMapping };
