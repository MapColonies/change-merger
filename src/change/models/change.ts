import { OsmChange } from '@map-colonies/node-osm-elements';

interface OsmXmlBase {
  id: number;
  version: number;
  changeset: number;
  tag?: OsmXmlTag | OsmXmlTag[];
}

interface ChangeActionObj {
  node: OsmXmlNode[];
  way: OsmXmlWay[];
}

interface NodeChange {
  node: OsmXmlNode | OsmXmlNode[];
}

interface WayChange {
  way: OsmXmlWay | OsmXmlWay[];
}

interface RelationChange {
  relation: unknown | unknown[];
}

export interface OsmXmlTag {
  k: string;
  v: string;
}

export interface OsmXmlNode extends OsmXmlBase {
  lon: number;
  lat: number;
}

export interface OsmXmlWay extends OsmXmlBase {
  nd: { ref: number }[];
}

export type ElementChange = NodeChange | WayChange | RelationChange;

export interface ChangeWithMetadata {
  externalId: string;
  change: OsmChange;
  tempOsmId?: number;
  action: 'create' | 'modify' | 'delete';
}

export interface MergedOsmXmlChange {
  version: '0.6';
  generator: string;
  create: ChangeActionObj;
  modify: ChangeActionObj;
  delete: ChangeActionObj;
}

export interface OsmXmlChange {
  version: '0.6';
  generator: string;
  create?: ElementChange | ElementChange[];
  modify?: ElementChange | ElementChange[];
  delete?: ElementChange | ElementChange[];
}
