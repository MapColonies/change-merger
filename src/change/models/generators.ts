import { OsmNode, OsmWay } from '@map-colonies/node-osm-elements';
import { OsmXmlNode, OsmXmlTag, OsmXmlWay } from './change';

const createOsmXmlTag = (oldTags?: Record<string, string>): OsmXmlTag[] => {
  if (!oldTags) {
    return [];
  }
  return Object.entries(oldTags).map(([k, v]) => ({ k, v }));
};

const createOsmXmlNode = (osmNode: OsmNode, changesetId: number): OsmXmlNode => {
  const { type, tags, ...node } = osmNode as Required<OsmNode>;
  return { ...node, changeset: changesetId, tag: createOsmXmlTag(tags) };
};

const createOsmXmlWay = (osmWay: OsmWay, tempMapping: Map<number, number>, changesetId: number): OsmXmlWay => {
  const { tags, nodes, type, ...rest } = osmWay as Required<OsmWay>;
  return {
    ...rest,
    changeset: changesetId,
    nd: nodes.map((node) => ({ ref: tempMapping.get(node.id) ?? node.id })),
    tag: createOsmXmlTag(tags),
  };
};

export { createOsmXmlNode, createOsmXmlWay, createOsmXmlTag };
