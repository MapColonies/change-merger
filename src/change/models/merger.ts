import { BaseElement, OsmNode, OsmWay } from '@map-colonies/node-osm-elements';
import { SERVICE_NAME } from '../../common/constants';
import { IdGenerator } from '../utils/idGenerator';
import { createOsmXmlNode, createOsmXmlWay } from './generators';
import { IdMapping, ChangeWithMetadata, OsmXmlChange, OsmXmlNode, OsmXmlWay } from './types';

const isNode = (element: BaseElement): element is OsmNode => element.type === 'node';
const isWay = (element: BaseElement): element is OsmWay => element.type === 'way';

const handleChangeArray = (
  elements: BaseElement[],
  idGenerator: IdGenerator,
  tempMapping: Map<number, number>,
  changesetId: number
): [OsmXmlNode[], OsmXmlWay?] => {
  const nodes: OsmXmlNode[] = [];
  let osmWay: OsmWay | null = null;

  for (let i = 0; i < elements.length; i++) {
    const element = elements[i];
    if (element.id < 0) {
      const newId = idGenerator.getId();
      tempMapping.set(element.id, newId);
      element.id = newId;
    }

    if (isNode(element)) {
      nodes.push(createOsmXmlNode(element, changesetId));
    } else if (isWay(element)) {
      osmWay = element;
    }
  }

  let xmlWay: OsmXmlWay | undefined = undefined;
  if (osmWay) {
    xmlWay = createOsmXmlWay(osmWay, tempMapping, changesetId);
  }
  return [nodes, xmlWay];
};

const handleChangeObj = (changeObj: ChangeWithMetadata, idGenerator: IdGenerator, change: OsmXmlChange, changesetId: number): number => {
  const tempMapping = new Map<number, number>();

  const actions: typeof changeObj.action[] = ['create', 'modify', 'delete'];

  actions.forEach((action) => {
    const elementArr = changeObj.change[action];
    if (elementArr) {
      const [nodes, way] = handleChangeArray(elementArr, idGenerator, tempMapping, changesetId);
      change[action].node.push(...nodes);
      if (way) {
        change[action].way.push(way);
      }
    }
  });

  return changeObj.tempOsmId !== undefined ? (tempMapping.get(changeObj.tempOsmId) as number) : 0;
};

const mergeChanges = (changes: ChangeWithMetadata[], changesetId: number): [OsmXmlChange, IdMapping[], string[]] => {
  const idGenerator = new IdGenerator();
  const change: OsmXmlChange = {
    generator: SERVICE_NAME,
    version: '0.6',
    create: { node: [], way: [] },
    modify: { node: [], way: [] },
    delete: { way: [], node: [] },
  };
  const idsToAdd: IdMapping[] = [];
  const idsToDelete: string[] = [];

  for (let i = 0; i < changes.length; i++) {
    const changeObj = changes[i];
    const tempOsmId = handleChangeObj(changeObj, idGenerator, change, changesetId);

    if (changeObj.action === 'create') {
      idsToAdd.push({ tempOsmId, externalId: changeObj.externalId });
    }

    if (changeObj.action === 'delete') {
      idsToDelete.push(changeObj.externalId);
    }
  }

  return [change, idsToAdd, idsToDelete];
};

export { mergeChanges };
