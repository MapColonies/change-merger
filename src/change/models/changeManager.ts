import { inject, injectable } from 'tsyringe';
import { Services } from '../../common/constants';
import { ILogger } from '../../common/interfaces';
import { changeToXml } from '../utils/jsonChangeToXml';
import { mergeChanges } from './merger';
import { IdMapping, ChangeWithMetadata } from './types';

@injectable()
export class ChangeManager {
  public constructor(@inject(Services.LOGGER) private readonly logger: ILogger) {}

  public mergeChanges(changes: ChangeWithMetadata[], changesetId: number): [string, IdMapping[], string[]] {
    this.logger.log('info', `merging ${changes.length} changes`);
    const [change, idsToCreate, idsToDelete] = mergeChanges(changes, changesetId);
    return [changeToXml({ osmChange: change }), idsToCreate, idsToDelete];
  }
}
