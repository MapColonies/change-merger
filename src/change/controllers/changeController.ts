import { RequestHandler } from 'express';
import httpStatus from 'http-status-codes';
import { injectable, inject } from 'tsyringe';
import { Services } from '../../common/constants';
import { ILogger } from '../../common/interfaces';

import { ChangeManager } from '../models/changeManager';
import { ChangeWithMetadata, IdMapping } from '../models/types';

interface ResponseChangeObject {
  change: string;
  created: IdMapping[];
  deleted: string[];
}

type MergeChangesHandler = RequestHandler<undefined, ResponseChangeObject, MergeChangesRequestBody>;

export interface MergeChangesRequestBody {
  changesetId: number;
  changes: ChangeWithMetadata[];
}

@injectable()
export class ChangeController {
  public constructor(@inject(Services.LOGGER) private readonly logger: ILogger, @inject(ChangeManager) private readonly manager: ChangeManager) {}

  public mergeChanges: MergeChangesHandler = (req, res) => {
    const [change, created, deleted] = this.manager.mergeChanges(req.body.changes, req.body.changesetId);
    return res.status(httpStatus.OK).json({ change, created, deleted });
  };
}
