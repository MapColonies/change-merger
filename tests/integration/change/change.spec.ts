import httpStatusCodes from 'http-status-codes';
import { container } from 'tsyringe';
import { MergeChangesRequestBody } from '../../../src/change/controllers/changeController';
import { RequestChangeObject } from '../../../src/change/models/types';
import { getSampleData } from '../../sampleData';

import { registerTestValues } from '../testContainerConfig';
import * as requestSender from './helpers/requestSender';

describe('resourceName', function () {
  beforeAll(function () {
    registerTestValues();
    requestSender.init();
  });
  afterEach(function () {
    container.clearInstances();
  });

  describe('Happy Path', function () {
    it('should return 200 status code and the resource', async function () {
      const requestBody: MergeChangesRequestBody = { changesetId: 3, changes: getSampleData() };

      const response = await requestSender.postMergeChanges(requestBody);

      expect(response.status).toBe(httpStatusCodes.OK);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe('Bad Path', function () {
    it('should fail if changesetId is not a number', async function () {
      const requestBody = ({ changesetId: 'aa', changes: getSampleData() } as unknown) as MergeChangesRequestBody;

      const response = await requestSender.postMergeChanges(requestBody);

      expect(response.status).toBe(httpStatusCodes.BAD_REQUEST);
      expect(response.body).toHaveProperty('message', 'request.body.changesetId should be number');
    });

    it('should fail if action value is not part of the enum', async function () {
      const changes = getSampleData();
      changes[0].action = 'xd' as 'create';
      const requestBody = ({ changesetId: 1, changes: changes } as unknown) as MergeChangesRequestBody;

      const response = await requestSender.postMergeChanges(requestBody);

      expect(response.status).toBe(httpStatusCodes.BAD_REQUEST);
      expect(response.body).toHaveProperty(
        'message',
        'request.body.changes[0].action should be equal to one of the allowed values: create, modify, delete'
      );
    });

    it('should fail if externalId is missing', async function () {
      const changes = getSampleData();
      const { externalId, ...newChange } = changes[0];
      changes[0] = newChange as RequestChangeObject;
      const requestBody = ({ changesetId: 1, changes: changes } as unknown) as MergeChangesRequestBody;

      const response = await requestSender.postMergeChanges(requestBody);

      expect(response.status).toBe(httpStatusCodes.BAD_REQUEST);
      expect(response.body).toHaveProperty('message', "request.body.changes[0] should have required property 'externalId'");
    });
  });
});
