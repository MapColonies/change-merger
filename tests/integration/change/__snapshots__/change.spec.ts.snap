// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change Happy Path should return 200 status code and the resource 1`] = `
Object {
  "change": "<osmChange generator=\\"change-merger\\" version=\\"0.6\\"><create><node id=\\"-1\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"3\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node><node id=\\"-2\\" version=\\"0\\" lon=\\"25\\" lat=\\"25\\" changeset=\\"3\\"></node><node id=\\"-3\\" version=\\"0\\" lon=\\"26\\" lat=\\"26\\" changeset=\\"3\\"></node><node id=\\"-5\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"3\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node><way id=\\"-4\\" version=\\"0\\" changeset=\\"3\\"><nd ref=\\"-1\\"></nd><nd ref=\\"-2\\"></nd><nd ref=\\"-3\\"></nd></way></create><modify><way id=\\"1\\" version=\\"0\\" changeset=\\"3\\"><nd ref=\\"-5\\"></nd><nd ref=\\"3\\"></nd><nd ref=\\"4\\"></nd></way></modify><delete><node id=\\"-6\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"3\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node></delete></osmChange>",
  "created": Array [
    Object {
      "externalId": "a",
      "tempOsmId": -4,
    },
  ],
  "deleted": Array [
    "c",
  ],
}
`;
